docs/syntax-multiple-index.md

```diff
@@ -0,0 +1,413 @@
+# Multiple indexing
+
+## Summary
+
+Introduce multiple indexing, an extension of table indexing that allows for multiple values to be read at the same time, with dedicated array and map shorthands for ergonomics.
+
+This allows for destructuring that:
+- works with array & map style tables
+- shortens both declaration and assignment
+- doesn't require parser backtrack
+- is consistent with Luau style
+
+## Motivation
+
+This is intended as a spiritual successor to the older ["Key destructuring" RFC by Kampfkarren](https://github.com/luau-lang/rfcs/pull/24), which was very popular but was unfortunately not able to survive implementation concerns.
+
+----
+
+Simple indexes on tables are very common both in and outside of Luau. A common use case is large libraries. It is common in the web world to see something like:
+
+```js
+const { useState, useEffect } = require("react");
+```
+
+...which allows you to quickly use `useState` and `useEffect` without fully qualifying it in the form of `React.useState` and `React.useEffect`. In Luau, if you do not want to fully qualify common React functions, the top of your file will often look like:
+
+```lua
+local useEffect = React.useEffect
+local useMemo = React.useMemo
+local useState = React.useState
+-- etc
+```
+
+...which creates a lot of redundant cruft.
+
+It is also common to want to have short identifiers to React properties, which basically always map onto a variable of the same name. As an anecdote, a regex search of `^\s+local (\w+) = \w+\.\1$` comes up 103 times in the My Movie codebase, many in the form of indexing React properties:
+
+```lua
+local position = props.position
+local style = props.style
+-- etc...
+```
+
+...whereas in JavaScript this would look like:
+```js
+const { position, style } = props
+
+// Supported in JavaScript, but not this proposal
+function MyComponent({
+	position,
+	style,
+})
+```
+
+React properties are themselves an example of a common idiom of passing around large tables as function arguments, such as with HTTP requests:
+
+```js
+// JavaScript
+get("/users", ({
+	users,
+	nextPageCursor,
+}) => { /* code */ })
+```
+
+## Design
+
+The `[]` indexing operator is extended to support explicitly reading from multiple (potentially) non-consecutive keys at once, by provided a comma-separated set of expressions.
+
+```Lua
+local numbers = {3, 5, 11}
+
+local three, five, eleven = numbers[1, 2, 3]
+```
+
+This will desugar to this behaviour exactly:
+
+```Lua
+local numbers = {3, 5, 11}
+local three, five, eleven = numbers[1], numbers[2], numbers[3]
```

Comment:
Really doesn't feel right to cause such a confusion with multi-dimensional array indexing for such a marginal feature.
(which might not exist today in Luau, but this will also closes the door on that)

Doesn't feel natural overall as well, but this part is my opinion.

Reply:
Got it, I'll look in other directions then :)